PHP: support per message compression disable

stanley-cheung opened the pull request:
* Fixes #2580
 * Needs stanley-cheung/Protobuf-PHP#2 for the codegen for client stub
 * C extension work: `OP_SEND_MESSAGE` now takes an array with key `array('message' => <a string>, 'flags' => <an array of options>)`, instead of just a string
 * `UnaryCall` and `ServerStreaming` `->start()` method now takes an additional `$options` array, which can carry the per-message compression disable write flags
 * `ClientStreaming` has an API change:
   * Instead of `start($arguments, $metadata)` which sends the metadata, goes through the array of `$arguments` and sends each of them, and closes the client, all together,
   * We now `start($metadata)` which just sends the metadata
   * Introduced a new method `write($data, $options)` which can be called by the client per message. This is also consistent with BidiStreaming
 * `BidiStreaming->write()` now takes an additional `$options` array
 * Updated unit tests, genereated code tests and interop tests

stanley-cheung commented:
@tbetbetbe @murgatroid99 Can you guys provide some feedback on this before I involve the API reviewers? Thanks!

tbetbetbe commented:
LGTM, please get API reviewers to take a look

murgatroid99 commented:
The general idea looks good to me, but it seems like all of the PHP code represents flags as an array, and the C code represents flags as a number and I don't see any code that turns it from one to the other. And why not just represent it as an integer all the way through?

stanley-cheung commented:
Right, I had a disconnect between my unit test and the actual library code. In my unit test, the flags is passed as a number but the library passsed them as an array. 

I would still like to have the API to take `$options` as an array for future expansion. But I will make sure the `flags` key is passed as a number. Sounds good?

murgatroid99 commented:
Yes, I think so. It sounds like `$options` will be an associative array, with `'flags'` as a key. That's what I would suggest, anyway.

stanley-cheung commented:
Updated the library to take `$options` as an associative array at the API level, with an expected key of `flags` holding a number. This gets passed down to the C extension layer. Also added a test at the API level to bridge the disconnect with unit tests. PTAL thanks!

murgatroid99 commented:
That part looks good to me now.

stanley-cheung commented:
Closing in favor of #3181 - need to merge into `release-0_11` beta branch instead
